Lesson 04/news_reader/index.js

```diff
@@ -0,0 +1,269 @@
+	let requestUrl = app.siteSettings.vesti.requestUrl;
+	
+	if (clientReq.body.newsCategory) {
+		requestUrl =
+			`${app.siteSettings.vesti.articleUrl}section.html?` +
+			`cid=${clientReq.body.newsCategory}`;
+	}
+	
+	request(requestUrl, (err, res, body) => {
+		const news = [];
+		const newsCount = parseInt(clientReq.body.newsCount);
+		let category;
+
+		if (!err && res.statusCode === 200) {
+			const $ = cheerio.load(body);
+			let $newsWrapper = '.news-wrapper_';
+			
+			if (clientReq.body.newsCategory) {
+				$newsWrapper = '.news-wrapper_section';
+				
+				app.siteSettings.vesti.categories.forEach((elem) => {
+					if (elem.cid === clientReq.body.newsCategory) {
+						category = elem.title;
+					}
+				});
+			}
+			
+			$('.b-item__inner', $newsWrapper).each((i, $elem) => {
+				// Skip exeeded amount of news items.
+				if (newsCount && i >= newsCount) {
+					return;
+				}
+				
+				const $title = $('.b-item__title a', $elem);
+
+				news.push({
+					num: i + 1,
+					title: $title.text().trim(),
+					text: $('.b-item__text', $elem).text().trim(),
+					time: $('.b-item__time', $elem).text().trim(),
+					url: $title.attr('href').trim().replace(/^\//, app.siteSettings.vesti.articleUrl)
+				});
+			});
+		}
+		
+		clientRes.render('news', {
+			title: app.viewSettings.pageTitle,
+			mainPage: app.viewSettings.mainPage,
+			newsSite: app.siteSettings.vesti.title,
+			newsCount: newsCount,
+			newsCategory: category,
+			news: news
+		});
+	});
+}
+
+// News handler.
+app.post('/news', urlencodedParser, (clientReq, clientRes) => {
+	switch (clientReq.body.newsSite) {
+		case 'vesti': fetchNewsVesti(clientReq, clientRes); break;
+	}
+});
+
+
+// News categories handler.
+app.get('/news/:site', (clientReq, clientRes) => {
+	const newsSite = app.siteSettings[clientReq.params.site];
+	
+	if (!newsSite) {
+		responseData.result = 404;
+		clientRes.send(JSON.stringify(responseData));
+		
+		return;
+	}
+	
+	request(newsSite.requestUrl, (err, res, body) => {
+		if (err) {
+			responseData.result = 1;
+			responseData.message = err.message;
+		} else if (res.statusCode !== 200) {
+			responseData.result = res.statusCode;
+			responseData.message = res.statusText;
+		}
+		
+		if (responseData.result) {
+			clientRes.send(JSON.stringify(responseData));
+			
+			return;
+		}
+		
+		switch (newsSite.name) {
+			case 'vesti': responseData.data = parseCategoryVesti(cheerio.load(body)); break;
+			default: break;
+		}
+
+		if (!responseData.data) {
+			responseData.result = 1;
+		}
+		
+		clientRes.send(JSON.stringify(responseData));
+	});
+});
+
+
+// Root request handler.
+app.get('/', (req, res) => {
+	const countOptions = [
+		{value: 10, text: '10'},
+		{value: 20, text: '20'},
+		{value: 30, text: '30'},
+		{value: 0, text: 'All'}
+	];
+	let selected;
+	
+	// Parse newsCount param.
+	if (req.query.newsCount) {
+		req.query.newsCount = parseInt(req.query.newsCount.trim());
+		
+		countOptions.forEach((elem) => {
```

Comment:
map

Reply:
Replaced with find() and cleaned up some "indian" code.